internal/parseutil/filereader.go

```diff
@@ -0,0 +1,56 @@
+package parseutil
+
+import (
+	"bytes"
+	"io"
+)
+
+// FileReader serves two purposes:
+// 1. read input sources and join them for single source
+// 2. track the lines counts of each input source
+type FileReader struct {
+	files  []file
+	source bytes.Buffer
+}
+
+type file struct {
+	path      string
+	startLine int // the number of first line of this file in joined source
+	lines     int
+}
+
+// AddFile adds input source
+func (fr *FileReader) AddFile(path string, source io.Reader) error {
+	curLen := fr.source.Len()
+	_, err := fr.source.ReadFrom(source)
+	if err != nil {
+		return err
+	}
+	if !bytes.HasSuffix(fr.source.Bytes(), []byte("\n")) {
+		fr.source.WriteByte('\n')
+	}
+	content := fr.source.Bytes()[curLen:]
+	lines := bytes.Count(content, []byte("\n"))
+	startLine := 1
+	if len(fr.files) > 0 {
```

Comment:
I think we should undo the "optimization" in https://github.com/benhoyt/goawk/pull/153/commits/00df7cd1443e48aac033b9f8cf9bea07c16f1c33 -- i.e., remove the `file.startLine` field. It complicates this code slightly (a few more lines and the `if` statement) and only saves a single integer addition in `FileLine`.

Reply:
Agree. Reverted.